### ISSUE TITLE
Enriching/PiggyBacking Information in ResilienceRetry

### ISSUE THREAD

**Issue by Kaustavd** [resilience4j/resilience4j#1908]
Resilience4j version:
1.3.1

Looks like resillience4j framework does not support retry call context to be enriched with some information piggybacking from previous attempt

Use case
 Http/RPC (Service A Calling Service B)  call failed due to timeout or has thrown non 2XX http response which caller application  aka Service A need to embed that signal in retry call to Service B

**Comment by RobWin** [comment]
Hi,
yes you are correct. It's not possible.

**Comment by Kaustavd** [comment]
@RobWin Is this not supported/possible because of functional programming nature or philosophical reasons (outside the scope of the responsibilities of the framework)?

**Comment by RobWin** [comment]
It's just not implemented. 
How would you like to access context information from a prev attemp?

**Comment by Kaustavd** [comment]
@RobWin Although in the production we are using `Retry.decorateSupplier` and have to check feasibility part. One of the ways to implement this in `Retry.decorateFunction` interface can accept another BiFunction  which can be invoked on exception block to enrich the requestContext.

**Comment by RobWin** [comment]
Hi, can you provide me a full example please?

**Comment by Kaustavd** [comment]
Current Interface looks like following
``` static <T, R> Function<T, R> decorateFunction(Retry retry, Function<T, R> function)``` 
where **function** is executed in infinite while loop which is terminated either on success or retry count exhaustion.

Proposed Interface
```    static  <T, R, E> Function<T, R> decorateFunction(Retry retry, Function<T, R> function, ThrowingBiFunction<T, R, RuntimeException> retryfunction) ```
where retryFunction will get executed on exception block or in successive retry count

**Comment by RobWin** [comment]
Can you show me an example how you want to enrich information from one request to another. I didn't understand it from your example.

**Comment by RobWin** [comment]
Hi,

have you implemented it in your application?
It should be possible without an additional method in the Retry interface. 
You can just copy `decorateFunction` and enhance it, or?

**Comment by Kaustavd** [comment]
I am not sure how same interface can work as ```decorateFunction``` interface allows  only pass one lambda to be excuted, as mentioned earlier in successive retry we need to execute it differently as previous execution context need to be incorporated. Hence the approach  I was thinking to pass ```retryFunction``` as extra argument which is nothing but a composition of previous lambda and execution context in the form of response (Happy Scenario) and exception (Unhappy Scenario) in the form of a biFunction

**Comment by RobWin** [comment]
I meant: copy the code of the decorateFunction and adapt it to your needs.

**Comment by Kaustavd** [comment]
Yes that is one way to do it. Is that the recommended way to do it? If that is the case I would raise PR.
One more thing is ```decorateFunction``` is little restrictive and does not suit if I have to decorate it along with circuitBreaker and Timelimiter. Sample code snippet which we are using in production
```Try.ofCallable(Timer.decorateCallable(clientTimer,
                            Timer.decorateCallable(serverTimer,
                                    CircuitBreaker.decorateCallable(circuitBreaker,
                                            TimeLimiter.decorateFutureSupplier(timeLimiter,
                                                    () -> threadPoolBulkhead.executeSupplier(Retry.decorateSupplier(retry,
                                                            () -> ThrowingSupplier.unchecked(throwingSupplier).get())).toCompletableFuture())))))
                    .get();
```

Could you please help if something of this sort possible to use with ```decorateFunction```

**Comment by Kaustavd** [comment]
Does that mean changes proposed above won't be merged back to master repo?

**Comment by Kaustavd** [comment]
Waiting for the update just in case you missed it
